oauth2_http/java/com/google/auth/oauth2/IamUtils.java

```diff
@@ -125,4 +128,67 @@ private static String getSignature(String serviceAccountEmail, Credentials crede
     GenericData responseData = response.parseAs(GenericData.class);
     return OAuth2Utils.validateString(responseData, "signedBlob", PARSE_ERROR_SIGNATURE);
   }
-}
+
+  /**
+   * Returns an IdToken issued to the serviceAccount with a specified
+   * targetAudience.
+   *
+   * @param serviceAccountEmail the email address for the service account to get
+   *                            an Id Token for
+   * @param credentials         credentials required for making the IAM call
+   * @param transport           transport used for building the HTTP request
+   * @param targetAudience      the audience the issued ID token should include
+   * @param additionalFields    additional fields to send in the IAM call
+   * @return New IdToken issed to the serviceAccount.
+   * @throws IOException if the IdToken cannot be issued.
+   * @see    https://cloud.google.com/iam/credentials/reference/rest/v1/projects.serviceAccounts/generateIdToken
+   */
+
+  static IdToken getIdToken(String serviceAccountEmail, Credentials credentials, HttpTransport transport,
+      String targetAudience, boolean includeEmail, Map<String, ?> additionalFields) throws IOException {
+
+    String idTokenUrl = String.format(ID_TOKEN_URL_FORMAT, serviceAccountEmail);
+    GenericUrl genericUrl = new GenericUrl(idTokenUrl);
+
+    GenericData idTokenRequest = new GenericData();
+    idTokenRequest.set("audience", targetAudience);
+    idTokenRequest.set("includeEmail", includeEmail);
+    for (Map.Entry<String, ?> entry : additionalFields.entrySet()) {
+      idTokenRequest.set(entry.getKey(), entry.getValue());
+    }
+    JsonHttpContent signContent = new JsonHttpContent(OAuth2Utils.JSON_FACTORY, idTokenRequest);
+
+    HttpCredentialsAdapter adapter = new HttpCredentialsAdapter(credentials);
+    HttpRequest request = transport.createRequestFactory(adapter).buildPostRequest(genericUrl, signContent);
+
+    JsonObjectParser parser = new JsonObjectParser(OAuth2Utils.JSON_FACTORY);
+    request.setParser(parser);
+    request.setThrowExceptionOnExecuteError(false);
+
+    HttpResponse response = request.execute();
+    int statusCode = response.getStatusCode();
+    if (statusCode >= 400 && statusCode < HttpStatusCodes.STATUS_CODE_SERVER_ERROR) {
+      GenericData responseError = response.parseAs(GenericData.class);
+      Map<String, Object> error = OAuth2Utils.validateMap(responseError, "error", PARSE_ERROR_MESSAGE);
+      String errorMessage = OAuth2Utils.validateString(error, "message", PARSE_ERROR_MESSAGE);
+      throw new IOException(String.format("Error code %s trying to getIDToken: %s", statusCode, errorMessage));
+    }
+    if (statusCode != HttpStatusCodes.STATUS_CODE_OK) {
```

Comment:
should this be any 200 level response?

Reply:
i'm not absolutely sure but i've never seen responses range 2xx  from this api layer:
 ref: https://cloud.google.com/apis/design/errors#handling_errors
 but storage mentions a bit of that here: https://cloud.google.com/storage/docs/json_api/v1/status-codes

(the google apis endpoint iamcredentials uses is older one (similar to drive api, calendar api, etc)